src/overmap.cpp

```diff
@@ -1352,139 +1389,95 @@ void overmap::draw(WINDOW *w, WINDOW *wbar, const tripoint &center,
                 // TODO: mission targets currently have no z-component, are assumed to be on z=0
                 // Mission target, display always, player should know where it is anyway.
                 ter_color = c_red;
-                ter_sym = '*';
-            } else if (blink && overmap_buffer.has_note(omx, omy, z)) {
+                ter_sym   = '*';
+            } else if (blink && overmap_buffer.has_note(cur_pos)) {
                 // Display notes in all situations, even when not seen
-                const std::string &note_text = overmap_buffer.note(omx, omy, z);
-
-                ter_color = c_yellow;
-                ter_sym = 'N';
-
-                int symbolIndex = note_text.find(':');
-                int colorIndex = note_text.find(';');
-
-                bool symbolFirst = symbolIndex < colorIndex;
-
-                if (colorIndex > -1 && symbolIndex > -1) {
-                    if (symbolFirst) {
-                        if (colorIndex > 4) {
-                            colorIndex = -1;
-                        }
-                        if (symbolIndex > 1) {
-                            symbolIndex = -1;
-                            colorIndex = -1;
-                        }
-                    } else {
-                        if (symbolIndex > 4) {
-                            symbolIndex = -1;
-                        }
-                        if (colorIndex > 2) {
-                            colorIndex = -1;
-                        }
-                    }
-                } else if (colorIndex > 2) {
-                    colorIndex = -1;
-                } else if (symbolIndex > 1) {
-                    symbolIndex = -1;
-                }
-
-                if (symbolIndex > -1) {
-
-                    int symbolStart = 0;
-                    if (colorIndex > -1 && !symbolFirst) {
-                        symbolStart = colorIndex + 1;
-                    }
-
-                    ter_sym = note_text.substr(symbolStart, symbolIndex - symbolStart).c_str()[0];
-
-                }
-
-                if (colorIndex > -1) {
-
-                    int colorStart = 0;
-                    if (symbolIndex > -1 && symbolFirst) {
-                        colorStart = symbolIndex + 1;
-                    }
-
-                    std::string sym = note_text.substr(colorStart, colorIndex - colorStart);
-
-                    ter_color = get_note_color( sym );
-                }
-
+                std::tie(ter_sym, ter_color, std::ignore) =
+                    get_note_display_info(overmap_buffer.note(cur_pos));
             } else if (!see) {
                 // All cases above ignore the seen-status,
                 ter_color = c_dkgray;
-                ter_sym = '#';
+                ter_sym   = '#';
                 // All cases below assume that see is true.
-            } else if (blink && npc_here) {
+            } else if (blink && overmap_buffer.has_npc(omx, omy, z)) {
```

Comment:
Ok wow, just this should help a ton, in my testing has_npc() was taking 40% of execution time.